Source/Fluids/WarpXFluidContainer.cpp

```diff
@@ -0,0 +1,743 @@
+/* Copyright 2023 Grant Johnson, Remi Lehe
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#include "ablastr/coarsen/sample.H"
+#include "Particles/Pusher/UpdateMomentumHigueraCary.H"
+
+#include "MusclHancockUtils.H"
+#include "WarpXFluidContainer.H"
+#include "WarpX.H"
+#include <ablastr/utils/Communication.H>
+using namespace ablastr::utils::communication;
+using namespace amrex;
+
+WarpXFluidContainer::WarpXFluidContainer(int nlevs_max, int ispecies, const std::string &name)
+{
+    species_id = ispecies;
+    species_name = name;
+
+    plasma_injector = std::make_unique<PlasmaInjector>(species_id, species_name);
+    physical_species = plasma_injector->getPhysicalSpecies();
+    charge = plasma_injector->getCharge();
+    mass = plasma_injector->getMass();
+
+    ReadParameters();
+
+    // Resize the list of MultiFabs for the right number of levels
+    N.resize(nlevs_max);
+    NU.resize(nlevs_max);
+}
+
+void WarpXFluidContainer::ReadParameters()
+{
+    static bool initialized = false;
+    if (!initialized)
+    {
+        const ParmParse pp_species_name(species_name);
+        pp_species_name.query("do_not_deposit", do_not_deposit);
+        pp_species_name.query("do_not_gather", do_not_gather);
+        pp_species_name.query("do_not_push", do_not_push);
+        initialized = true;
+    }
+}
+
+void WarpXFluidContainer::AllocateLevelMFs(int lev, const BoxArray &ba, const DistributionMapping &dm)
+{
+    int ncomps = 1;
+    amrex::IntVect nguards = {AMREX_D_DECL(2, 2, 2)};
+
+    auto &warpx = WarpX::GetInstance();
+
+    // set human-readable tag for each MultiFab
+    auto const tag = [lev](std::string tagname)
+    {
+        tagname.append("[l=").append(std::to_string(lev)).append("]");
+        return tagname;
+    };
+
+    warpx.AllocInitMultiFab(N[lev], amrex::convert(ba, amrex::IntVect::TheNodeVector()),
+                            dm, ncomps, nguards, tag("fluid density"), 0.0_rt);
+
```

Comment:
We would need to allocate these arrays with proper staggering:
```
    // Temporary Half-step values
    amrex::BoxArray ba = N[lev]->boxArray();
    amrex::MultiFab tmp_Q_minus_x( amrex::convert(ba, IntVect(0,1,1)), N[lev]->DistributionMap(), 4, 1);
    amrex::MultiFab tmp_Q_plus_x( amrex::convert(ba, IntVect(0,1,1)), N[lev]->DistributionMap(), 4, 1);
    amrex::MultiFab tmp_Q_minus_y( amrex::convert(ba, IntVect(1,0,1)), N[lev]->DistributionMap(), 4, 1);
    amrex::MultiFab tmp_Q_plus_y( amrex::convert(ba, IntVect(1,0,1)), N[lev]->DistributionMap(), 4, 1);
    amrex::MultiFab tmp_Q_minus_z( amrex::convert(ba, IntVect(1,1,0)), N[lev]->DistributionMap(), 4, 1);
    amrex::MultiFab tmp_Q_plus_z( amrex::convert(ba, IntVect(1,1,0)), N[lev]->DistributionMap(), 4, 1);
```
But then this would probably require some changes in the calculation of indices.